src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/AutoInsert/RemoteAutoInsertService.cs

```diff
@@ -0,0 +1,218 @@
+    }
+
+    private readonly IAutoInsertService _autoInsertService
+        = args.ExportProvider.GetExportedValue<IAutoInsertService>();
+    private readonly IDocumentMappingService _documentMappingService
+        = args.ExportProvider.GetExportedValue<IDocumentMappingService>();
+    private readonly IFilePathService _filePathService =
+        args.ExportProvider.GetExportedValue<IFilePathService>();
+    private readonly IRazorFormattingService _razorFormattingService =
+        args.ExportProvider.GetExportedValue<IRazorFormattingService>();
+    private readonly ILogger _logger =
+        args.ExportProvider.GetExportedValue<ILoggerFactory>()
+        .GetOrCreateLogger(nameof(RemoteAutoInsertService));
+
+    public ValueTask<Response> TryResolveInsertionAsync(
+        RazorPinnedSolutionInfoWrapper solutionInfo,
+        DocumentId documentId,
+        LinePosition linePosition,
+        string character,
+        bool autoCloseTags,
+        bool formatOnType,
+        bool indentWithTabs,
+        int indentSize,
+        CancellationToken cancellationToken)
+        => RunServiceAsync(
+            solutionInfo,
+            documentId,
+            context => TryResolveInsertionAsync(
+                context,
+                linePosition,
+                character,
+                autoCloseTags,
+                formatOnType,
+                indentWithTabs,
+                indentSize,
+                cancellationToken),
+            cancellationToken);
+
+    private async ValueTask<Response> TryResolveInsertionAsync(
+        RemoteDocumentContext remoteDocumentContext,
+        LinePosition linePosition,
+        string character,
+        bool autoCloseTags,
+        bool formatOnType,
+        bool indentWithTabs,
+        int indentSize,
+        CancellationToken cancellationToken)
+    {
+        var sourceText = await remoteDocumentContext.GetSourceTextAsync(cancellationToken).ConfigureAwait(false);
+        if (!sourceText.TryGetAbsoluteIndex(linePosition, out var index))
+        {
+            return Response.NoFurtherHandling;
+        }
+
+        var codeDocument = await remoteDocumentContext.GetCodeDocumentAsync(cancellationToken).ConfigureAwait(false);
+
+        // Always try our own service first, regardless of language
+        // E.g. if ">" is typed for html tag, it's actually our auto-insert provider
+        // that adds closing tag instead of HTML even though we are in HTML
+        var insertTextEdit = _autoInsertService.TryResolveInsertion(
+            codeDocument,
+            VsLspExtensions.ToPosition(linePosition),
+            character,
+            autoCloseTags);
+
+        if (insertTextEdit is { } edit)
+        {
+            return Response.Results(RemoteInsertTextEdit.FromLspInsertTextEdit(edit));
+        }
+
+        var positionInfo = PreferHtmlInAttributeValuesDocumentPositionInfoStrategy.Instance.GetPositionInfo(_documentMappingService, codeDocument, index);
```

Comment:
It seems odd that we're using the position info strategy only after we've already tried to respond with our service. Doesn't that mean we could be making decisions based on the wrong language? The existing endpoint always gets the position info first.

Reply:
No, the new code does what the existing code did. Existing code does get position info right away, but

1. It always calls "our" (Razor) handler first before trying to delegate

[https://github.com/dotnet/razor/blob/3423142c7dc64c6ef1676da7e7c5439a7962e222/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/AbstractRazorDelegatingEndpoint.cs#L124](https://github.com/dotnet/razor/blob/3423142c7dc64c6ef1676da7e7c5439a7962e222/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/AbstractRazorDelegatingEndpoint.cs#L124)

2. In OnAutoInsert specifically, our service never looks at the passed in position info, instead using position from request

[https://github.com/dotnet/razor/blob/3423142c7dc64c6ef1676da7e7c5439a7962e222/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/AutoInsert/OnAutoInsertEndpoint.cs#L104](https://github.com/dotnet/razor/blob/3423142c7dc64c6ef1676da7e7c5439a7962e222/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/AutoInsert/OnAutoInsertEndpoint.cs#L104)